pkg/tenancy/write.go

```diff
@@ -0,0 +1,76 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package tenancy
+
+import (
+	"fmt"
+
+	"github.com/timescale/promscale/pkg/prompb"
+)
+
+// writeAuthorizer is a write authorizer that authorizes if the incoming write request is valid to be written or not.
+type writeAuthorizer struct {
+	AuthConfig
+}
+
+var errTenantMismatch = fmt.Errorf("__tenant__ value and tenant-name from headers are different")
+
+// NewAuthorizer returns a new plainWriteAuthorizer.
+func NewWriteAuthorizer(config AuthConfig) WriteAuthorizer {
+	return &writeAuthorizer{config}
+}
+
+func (a *writeAuthorizer) isAuthorized(tenantName string) (err error, isNonTenant bool) {
+	if tenantName == "" {
+		if a.allowNonTenants() {
+			// Non-multi-tenant write request.
+			// We accept non-multi-tenant write requests in tenancy mode.
+			return nil, true
+		} else {
+			return fmt.Errorf("non multi-tenant writes not allowed"), true
+		}
+	}
+	if a.IsTenantAllowed(tenantName) {
+		return nil, false
+	}
+	return fmt.Errorf("%s %s", ErrUnauthorizedTenant.Error(), tenantName), false
+}
+
+func (a *writeAuthorizer) VerifyAndApplyTenantLabel(tenantNameFromHeader string, labels []prompb.Label) ([]prompb.Label, error) {
```

Comment:
We talked on slack about re-writing a bunch of the logic here.

Reply:
Just looking at the function name, it seems like it needs to be split up into two since its clearly breaking SRP.

Reply:
after the suggestion which mat gave, it should solve all the concerns here.